enhancements/network/ovn-kubernetes-ipsec.md

```diff
@@ -0,0 +1,295 @@
+	// HybridOverlayConfig configures an additional overlay network for peers that are
+	// not using OVN.
+	// +optional
+  HybridOverlayConfig *HybridOverlayConfig `json:"hybridOverlayConfig,omitempty"`
+  // EnableIPsec enables IPsec encryption on all pod-pod traffic within the cluster. If unset, the plugin will default
+  // EnableIPsec to false.
+  // +optional
+  EnableIPsec *bool `json:"enableIPsec,omitempty"`
+}
+```
+
+This will result in an updated OVNKubernetes Operator configuration object:
+(.spec.defaultNetwork.ovnKubernetesConfig) which will enable cluster-wide
+configuration of IPsec at cluster installation time:
+
+```
+spec:
+  defaultNetwork:
+    type: OVNKubernetes
+    ovnKubernetesConfig:
+      mtu: 1400
+      genevePort: 6081
+      enableIPsec: true/false
+```
+
+### Implementation Details/Notes/Constraints
+
+It is expected that the following changes will need to be made in order to
+enable this enhancement:
+
+- OVNKubernetes pod Dockerfiles will need to be updated in order to install
+  necessary components required by OVN/OVS to enable IPsec (e.g.
+  openvswitch-ipsec, libreswan, etc).
+- Modify Cluster Network Operator manifests to configure start up of packages
+  required by OVN/OVS in order to enable IPsec (e.g. ovs-monitor-ipsec, pluto,
+  etc) and determine how to containerize these components.
+- Ensure that all versions of all components are at the correct level required
+  by OVN/OVS. This may required back-porting of patches into RHEL. It has
+  already been identified that OVS and the Kernel will each require a patch to
+  enable this enhancement.
+- Ensure correct generation of the Certificate Authority (CA) certificate and
+  all node private keys and CA-signed certificates.
+- Modify Cluster Network Operator manifests to ensure that all OVS/OVN
+  components are configured to use host private key, host certificate and CA
+  certificate.
+- Modify OVNKubernetes operator configuration object in order to allow
+  enablement of IPsec for the cluster.
+- Certificate rotation will be handled in the same as it is with the current
+  OVN TLS implementation.
+
+### Risks and Mitigations
+
+#### Upstreaming
+
+A number of changes will need to be made to upstream communities including the
+Linux kernel. As these may need time to execute, it is imperative that we
+identify them early and start the process.
+
+#### Performance
+
+Enablement of IPsec will cause degradation of overall networking throughput and
+latency due to the additional overhead of the IPsec protocol. Some early
+performance testing will be done in a non-production cluster in order to get
+an approximation of the expected performance degradation. These results will be
+posted at a later date. After implementation of the enhancement, it is expected
+that comprehensive benchmarking and profiling will take place. Based on the
+output of this activity, it may be decided to start some optimization activities
+in parallel to improve performance.
```

Comment:
Is it worth testing on an Intel NIC (82599 aka. x540) with hardware offload, then the overhead should be small? They're pretty common and cheap. Oracle originally added support - https://blogs.oracle.com/linux/tips-and-tricks-for-ipsec-on-intel-10-gbe-nics-v2

Reply:
Given software IPsec is often limited to <1Gbps, wouldn't most serious use _require_ hardware offload? Looking at BZ #1736260 offload sometimes requires transport mode rather than the default tunnel mode.

Reply:
You could be right but the focus of the initial implementation wont be HW offload and performance. If I get a chance I will try and test it and it certainly will be a focus of later enhancements.

This implementation will use transport mode.